nmt.go

```diff
@@ -0,0 +1,240 @@
+// Package nmt contains an NMT implementation.
+package nmt
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/liamsi/merkletree"
+
+	"github.com/lazyledger/nmt/internal"
+	"github.com/lazyledger/nmt/namespace"
+)
+
+var (
+	ErrMismatchedNamespaceSize = errors.New("mismatching namespace sizes")
+	ErrInvalidPushOrder        = errors.New("pushed data has to be lexicographically ordered by namespace IDs")
+)
+
+type NamespacedMerkleTree struct {
+	treeHasher Hasher
+	tree       *merkletree.Tree
+
+	// just cache stuff until we pass in a store and keep all nodes in there
+	leaves     []namespace.PrefixedData
+	leafHashes [][]byte
+	// this can be used to efficiently lookup the range for an
+	// existing namespace without iterating through the leaves
+	namespaceRanges map[string]merkletree.LeafRange
+	minNID          namespace.ID
+	maxNID          namespace.ID
+}
+
+func New(treeHasher Hasher) *NamespacedMerkleTree {
+	return &NamespacedMerkleTree{
+		treeHasher: treeHasher,
+		tree:       merkletree.NewFromTreehasher(treeHasher),
+		// XXX: 100 seems like a good capacity for the leaves slice
```

Comment:
What are the consequences of this parameter? Does it mean that only up to 100 leafs can be pushed?

Reply:
This only determines the initial capacity. Meaning up to 100 items should be appendable without an additional mem allocation. `append`s after that initial capacity will increase the underlying capacity s.t. the append will still work (the capacity will roughly be doubled as needed: https://github.com/golang/go/blob/ba9e10889976025ee1d027db6b1cad383ec56de8/src/runtime/slice.go#L125).

Reply:
Given that the number of shares will always be a power of 2 for LazyLedger, would it not be better to set this initial capacity to a power of 2 as well? An original square with size 64*64 (extended 128*128) would give ~1MB blocks, so 128 here seems like a reasonable parameter choice for initial capacity to avoid having to always do a re-size operation.

Reply:
Later, maybe we should expose this as a param to the constructor. e.g. in LL the caller will know the exact capacity and if that's passed in, only one allocation will be necessary. This should not matter if the allocations happen on the stack only. I never fully understood which stuff ends up on the heap vs stack in golang but apparently this allocation is on the heap (which is more expensive). Check for yourself via: `go build -gcflags -m`
> [...]
> ` make([][]byte, 0, 100) escapes to heap`
> [...]

Reply:
> so 128 here seems like a reasonable parameter choice for initial capacity 

I agree.

Reply:
Changed to 128 in https://github.com/lazyledger/nmt/pull/3/commits/056112da486a306f9e88ac60f306025a9cebb9ed

We might want to consider the following options (potentially in followup PRs):
 - make this a param in the constructor
   - pros: full control / can be optimized by use-case (small vs big trees)
   - cons: leaky abstraction? (leaks the internals of the implementation in some sense)
 - make the constructor take in all the leafs at once :
   - pros: optimal with regards to allocs & does not leak any internals; might be easier to read as you can pass in the whole data at once
   - cons: ?

Reply:
Perhaps there can be a method to change this parameter, instead of putting in the constructor and leaking the abstraction?